# Ficha de trabalho 4

1- Considera o dataset cachexia. Lê os dados usando a função read_csv do pandas (dados e metadados).

In [2]:
import pandas as pd

data = pd.read_csv("data_cachexia.csv", index_col=0)
meta = pd.read_csv("meta_cachexia.csv", index_col=0)
print(data.shape)
print(meta.shape)

(77, 63)
(77, 1)


2 -Aplica a transformação necessária para que os dados seja escalonados para ter média 0 e desvio padrão 1. Verifica que as médias de todas as colunas é aproximadamente zero.

In [3]:
from sklearn import preprocessing

data_scaled = preprocessing.scale(data.loc[:, data.columns != "Patient ID"])
data_scaled

print("Data mean:", data_scaled.mean())
print("Standard deviation: ", data_scaled.std())

Data mean: -1.8675365658506033e-17
Standard deviation:  1.0


3 - Cria dois conjuntos de dados treino e teste para serem usados na criação e validação de modelos de aprendizagem máquina. Considera 30% das amostras para formar o conjunto de teste. 

In [4]:
import numpy as np


input_data = data.values
output_data = meta.values[:, 0]

number_samples = int(data_scaled.shape[0] * 0.3)

indexes = np.random.permutation(len(input_data))

train_input = data_scaled[indexes[:-number_samples]]
train_output = output_data[indexes[:-number_samples]]

test_input = data_scaled[indexes[-number_samples:]]
test_output = output_data[indexes[-number_samples:]]



4 - Constroi diversos tipos de modelos de classificação treinando e testando com os conjuntos criados anteriormente. Calcula, para cada um dos modelos no test set, as métricas de erro PECC e F1-score, e calcula a matriz de confusão. Considere como classe positiva a classe "cachexic"

In [5]:
# function to calculate the PECC error

def pecc_error (predicted, real):
    return sum([(predicted[i] == real[i]) for i in range(len(predicted))]) / len(predicted)

# f1 score import
from sklearn.metrics import f1_score

In [11]:
# KNNClassifier model

from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier()
knn_model.fit(train_input, train_output)
knn_predict = knn_model.predict(test_input)

print("Score: ", knn_model.score(test_input, test_output))
print("PECC error: ", pecc_error(knn_predict, test_output))
print("f1-score: ", f1_score(test_output, knn_predict, average="macro"))
print("\n", "Predicted values \n", knn_predict, "\n", "Real values \n",  test_output)

Score:  0.5652173913043478
PECC error:  0.5652173913043478
f1-score:  0.5436507936507937

 Predicted values 
 ['control' 'control' 'control' 'cachexic' 'control' 'cachexic' 'cachexic'
 'control' 'control' 'cachexic' 'cachexic' 'cachexic' 'control' 'cachexic'
 'control' 'cachexic' 'control' 'cachexic' 'cachexic' 'cachexic'
 'cachexic' 'cachexic' 'cachexic'] 
 Real values 
 ['control' 'control' 'cachexic' 'control' 'cachexic' 'cachexic' 'cachexic'
 'cachexic' 'control' 'cachexic' 'cachexic' 'control' 'cachexic' 'control'
 'control' 'cachexic' 'cachexic' 'cachexic' 'cachexic' 'control'
 'cachexic' 'cachexic' 'control']


In [10]:
# Naive bayes model

from sklearn.naive_bayes import GaussianNB

gnb_model = GaussianNB()
gnb_model.fit(train_input, train_output)
gnb_predict = gnb_model.predict(test_input)

print("Score: ", gnb_model.score(test_input, test_output))
print("PECC error: ", pecc_error(gnb_predict, test_output))
print("f1-score: ", f1_score(test_output, gnb_predict, average="macro"))
print("\n", "Predicted values \n", gnb_predict, "\n", "Real values \n", test_output)

Score:  0.5217391304347826
PECC error:  0.5217391304347826
f1-score:  0.5180952380952382

 Predicted values 
 ['control' 'control' 'control' 'cachexic' 'cachexic' 'cachexic' 'cachexic'
 'control' 'control' 'cachexic' 'control' 'cachexic' 'control' 'control'
 'control' 'control' 'control' 'cachexic' 'cachexic' 'cachexic' 'cachexic'
 'control' 'cachexic'] 
 Real values 
 ['control' 'control' 'cachexic' 'control' 'cachexic' 'cachexic' 'cachexic'
 'cachexic' 'control' 'cachexic' 'cachexic' 'control' 'cachexic' 'control'
 'control' 'cachexic' 'cachexic' 'cachexic' 'cachexic' 'control'
 'cachexic' 'cachexic' 'control']


In [12]:
# Decision Tree model

from sklearn.tree import DecisionTreeClassifier

tree_model = DecisionTreeClassifier()
tree_model.fit(train_input, train_output)
tree_predict = tree_model.predict(test_input)

print("Score: ", tree_model.score(test_input, test_output))
print("PECC error: ", pecc_error(tree_predict, test_output))
print("f1_score: ", f1_score(test_output, tree_predict, average="macro"))
print("\n", "Predicted values \n", tree_predict, "\n", "Real values \n",  test_output)

Score:  0.6956521739130435
PECC error:  0.6956521739130435
f1_score:  0.6537634408602151

 Predicted values 
 ['control' 'control' 'cachexic' 'cachexic' 'cachexic' 'cachexic'
 'cachexic' 'control' 'control' 'cachexic' 'cachexic' 'cachexic'
 'cachexic' 'control' 'cachexic' 'cachexic' 'cachexic' 'cachexic'
 'cachexic' 'cachexic' 'cachexic' 'control' 'cachexic'] 
 Real values 
 ['control' 'control' 'cachexic' 'control' 'cachexic' 'cachexic' 'cachexic'
 'cachexic' 'control' 'cachexic' 'cachexic' 'control' 'cachexic' 'control'
 'control' 'cachexic' 'cachexic' 'cachexic' 'cachexic' 'control'
 'cachexic' 'cachexic' 'control']


In [9]:
# confusion matrixes
from sklearn.metrics import confusion_matrix

# rows represent true values (test_output)
# columns represent predicted values
print("KNN CONFUSION MATRIX \n", confusion_matrix(test_output, knn_predict))
print("NAIVE BAYES CONFUSION MATRIX \n", confusion_matrix(test_output, gnb_predict))
print("DECISION TREE CONFUSION MATRIX \n", confusion_matrix(test_output, tree_predict))

KNN CONFUSION MATRIX 
 [[9 5]
 [5 4]]
NAIVE BAYES CONFUSION MATRIX 
 [[7 7]
 [4 5]]
DECISION TREE CONFUSION MATRIX 
 [[10  4]
 [ 3  6]]


5 - Usando apenas o melhor modelo do exercício anterior, compare o seu desempenho considerando dados com transformação logaritmica e posteriormente standardizados

In [14]:
log_data = np.log2(data)

log_input_data = log_data.values
output_data = meta.values[:, 0]

log_input_scaled = preprocessing.scale(log_input_data)

print("Mean: ", log_input_scaled.mean())
print("Standard deviation: ", log_input_scaled.std())

Mean:  5.163189329116374e-17
Standard deviation:  1.0


In [19]:
number_samples = int(log_input_scaled.shape[0] * 0.3)

indexes = np.random.permutation(len(log_input_scaled))

log_train_input = log_input_scaled[indexes[:-number_samples]]
log_train_output = output_data[indexes[:-number_samples]]

log_test_input = log_input_scaled[indexes[-number_samples:]]
log_test_output = output_data[indexes[-number_samples:]]

log_train_input.shape, log_train_output.shape, log_test_input.shape, log_test_output.shape


((54, 63), (54,), (23, 63), (23,))

In [21]:
from sklearn.tree import DecisionTreeClassifier

tree_log_model = DecisionTreeClassifier()
tree_log_model.fit(log_train_input, log_train_output)
tree_log_predict = tree_log_model.predict(log_test_input)

print("\n", "Predicted values \n", tree_log_predict, "\n", "Real values \n",  log_test_output)
print("Score: ", tree_log_model.score(log_test_input, log_test_output))


 Predicted values 
 ['control' 'cachexic' 'cachexic' 'cachexic' 'control' 'cachexic'
 'cachexic' 'control' 'control' 'cachexic' 'control' 'control' 'control'
 'cachexic' 'cachexic' 'cachexic' 'cachexic' 'cachexic' 'control'
 'cachexic' 'control' 'cachexic' 'cachexic'] 
 Real values 
 ['cachexic' 'control' 'cachexic' 'cachexic' 'cachexic' 'cachexic'
 'cachexic' 'cachexic' 'cachexic' 'control' 'control' 'cachexic' 'control'
 'control' 'control' 'cachexic' 'cachexic' 'cachexic' 'cachexic'
 'cachexic' 'cachexic' 'cachexic' 'cachexic']
Score:  0.5217391304347826


With a score of 0.52, considering that the best possible score is 1, our decision tree model is actually average for this problem